### ISSUE TITLE
flatbuffer.shape() might return nullptr

### ISSUE THREAD

**Issue by dauras** [tensorflow/tensorflow#38039]
For certain tensors (e.g. multiply with a scalar (constant) tensor, such as 1/6 at the output of hard sigmoid), the shape attribute might be None/undefined and hence, flatbuffer.shape() returns a nullptr.

In this code
https://github.com/tensorflow/tensorflow/blob/59c06b9016700dbf1ab0cefc062d247345cdd0f0/tensorflow/lite/micro/memory_helpers.cc#L82
this nullptr is dereferenced unconditionally.

**Comment by gadagashwini** [comment]
@dauras , Please provide the standalone code to reproduce the issue. Thanks

**Comment by dauras** [comment]
Sorry, I do not have an example at hand, but I can show you that TFL handles the shape==nullptr case differently than TFLM:

https://github.com/tensorflow/tensorflow/blob/9fdaa925f4825b6c00da6f13fda7d75522fb82d6/tensorflow/lite/interpreter_builder.cc#L184-L188

https://github.com/tensorflow/tensorflow/blob/9fdaa925f4825b6c00da6f13fda7d75522fb82d6/tensorflow/lite/kernels/internal/tensor_ctypes.h#L34-L36

So, shape==nullptr is converted to an empty STL vector, or an empty RuntimeShape. Those are then similar/identical to [1,1,1,1], for where it matters (e.g. broadcast operators or buffer size calculations).

**Comment by dauras** [comment]
I found another motivating example:

- Download [MobileNet v3 small checkpoint](https://storage.googleapis.com/mobilenet_v3/checkpoints/v3-small_224_1.0_uint8.tgz)
- Extract the .tflite file
- Decode with flatc
      `flatc -t --defaults-json tensorflow/lite/schema/schema.fbs -- v3-small_224_1.0_uint8.tflite`
- Inspect `v3-small_224_1.0_uint8.json`
  Look for tensor named `MobilenetV3/expanded_conv_10/squeeze_excite/Conv_1/mul/y`
  It has an empty shape:

```
        {       
          shape: [

          ],      
          type: "UINT8",
          buffer: 151,
          name: "MobilenetV3/expanded_conv_10/squeeze_excite/Conv_1/mul/y",
          quantization: {
            min: [  
              0.0     
            ],      
            max: [  
              0.16667 
            ],      
            scale: [
              0.000654
            ],      
            zero_point: [
              0       
            ],      
            details_type: 0,
            quantized_dimension: 0
          },      
          is_variable: false
        },  
```

**Comment by wangtz** [comment]
Ah that's an interesting point. If the shape is not known in advance (nullptr in the flatbuffer), TFLM interpreter won't be able to pre-calculate the memory usage for allocation. TFLite workaround this limitation by running prepare method to re-calculate the output shape for each layer, which is not support in TFLM.    I would suggest to investigate why the shape is undermined in the model.

**Comment by dauras** [comment]
This tensor is one of the operands of a multiplication operator. This multiplication operator is part of the hard-sigmoid activation function ``tf.nn.relu6(x+3)*0.16667``, used in MobileNet v3. The tensor is a scalar, and it is constant. Its value is ``0.16667``.
https://github.com/tensorflow/models/blob/295b3b4f2d304a6e90b518755fd6d5bf90005975/research/slim/nets/mobilenet/mobilenet_v3.py#L40

**Comment by wangtz** [comment]
I see. You meant that BytesRequiredForTensor should be able to handle scalar. I think that might be reasonable. We haven't had a case for passing scalar as a tensor. But feel free to submit a PR if you're blocked on this.

**Comment by oleksii-abramenko** [comment]
Confirm the issue. It can be reproduced in the model with GRU when performing OPTIMIZE_FOR_SIZE quantization. In order to observe the issue one should also provide converter.representative_dataset, to enforce quantization of activations.

Dereferencing of the NULL pointer happens when processing tensors  **model/gru/gru_cell_3/split/split_dim** and  **model/gru/gru_cell_3/sub/x**.

Can anyone suggest a workaround for this?

**Comment by tilakrayal** [comment]
@dauras,
The file **memory_helpers.cc** was updated with the `flatbuffer_tensor.shape == nullptr`.

https://github.com/tensorflow/tflite-micro/blob/main/tensorflow/lite/micro/memory_helpers.cc#L104

```
TfLiteStatus BytesRequiredForTensor(const tflite::Tensor& flatbuffer_tensor,
                                    size_t* bytes, size_t* type_size) {
  int element_count = 1;
  // If flatbuffer_tensor.shape == nullptr, then flatbuffer_tensor is a scalar
  // so has 1 element.
  if (flatbuffer_tensor.shape() != nullptr) {
    for (size_t n = 0; n < flatbuffer_tensor.shape()->Length(); ++n) {
      element_count *= flatbuffer_tensor.shape()->Get(n);
    }
  }
```

**Comment by mihaimaruseac** [comment]
Awesome. In this case, this can be closed.
